### PR TITLE
feat(ai): First implementation of TritonServer metrics feature

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
This PR adds a new API for managing status and performance metrics of an Inference Engine. The implementation for the Triton Server is added as well.

**Related Issue:** This PR fixes/closes N/A

**Description of the solution adopted:** The new `InferenceEngineMetricsService` interface provides APIs to get metrics from an inference engine in the form of a Map whose keys are the metric names. 

The implementation on the TritonServer engine, retrieves the GPU metrics from the metrics ports via http and the model statistics from the inference port via grpc. Both metrics are parsed and re-arranged in a json format. The keys for the GPU metrics are:

```
gpu_metrics.<gpu_uuid>
```

where the `gpu_uuid` is an identifier for the GPU device. For the model statistics, instead, the keys are:

```
model_metrics.<model.name>.<version>
```

Moreover, the old TritonServer implementation has been deleted, since it has been deprecated in previous releases.

### CODE REVIEW

**Comment by pierantoniomerlino** [comment]
@mattdibi could you please take a look at this PR, even if it is still in draft? I've to test some corner cases and check some code choices, but the main stuff is here.

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 1]
```diff
@@ -158,6 +158,51 @@
...
+        
+        <AD id="enable.cpu.metrics"
+            name="Enable Triton Server CPU Metrics"
+            type="Boolean"
+            cardinality="0"
+            required="true"
+            default="true"
+            description="Enable the Triton Server CPU Metrics. It requires that the Enable Triton Server Metrics property is set to true.">
+        </AD>      
+
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]: 🤔  I'm worried about the exploding number of configuration parameters. Can we take some decisions for our users? What if `enable.metrics` enable both `cpu` and `gpu`? Is that feasible?

```suggestion
        <AD id="enable.metrics"
            name="Enable Triton Server Metrics"
            type="Boolean"
            cardinality="0"
            required="true"
            default="true"
            description="Enable the Triton Server Metrics feature. This property enables all the default metrics..">
        </AD>  
        
```

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/MetricsParser.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * 
+     * <metric_name>-<model-name>-<parameter-1>-<parameter-2>...
+     * 
+     * For example the metric
+     * 
+     * nv_inference_request_failure{model="preprocessor",reason="OTHER",version="1"} 0
+     * 
+     * is converted to
+     * 
+     * nv_inference_request_failure-preprocessor-OTHER-1 0
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/MetricsParser.java]: Is there a reason why we need to convert the metrics this way? (thinking about it it's probably the `{` right?)... as a user I would prefer to only learn one syntax. Having a middle man (i.e. Kura in this case) changing the metrics syntax arbitrarily is really annoying.

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java [hunk 3]
```diff
@@ -264,6 +264,16 @@ private ContainerConfiguration createContainerConfiguration() {
         if (!this.options.getBackendsConfigs().isEmpty()) {
             this.options.getBackendsConfigs().forEach(config -> entrypointOverride.add("--backend-config=" + config));
         }
+        entrypointOverride.add("--allow-metrics=" + this.options.areMetricsEnabled());
+        entrypointOverride.add("--allow-gpu-metrics=" + this.options.areGpuMetricsEnabled());
+        entrypointOverride.add("--allow-cpu-metrics=" + this.options.areCpuMetricsEnabled());
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java]: As I said above I would try to reduce the number of user configurable parameters:

```suggestion
        entrypointOverride.add("--allow-metrics=" + this.options.areMetricsEnabled());
        entrypointOverride.add("--allow-gpu-metrics=" + this.options.areMetricsEnabled());
        entrypointOverride.add("--allow-cpu-metrics=" + this.options.areMetricsEnabled());
```

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml (hunk 1 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]: I'm worried too. Currently we have the following parameters:

- default metrics
- GPU metrics
- CPU metrics
- metrics config

I'd preserve the metrics enable and config, setting GPU and CPU enabled by default.

**Comment by pierantoniomerlino** [comment]
Tested on a Nvidia Jetson Orin Nano with the following scenarios:

- containerized TritonServer on the Orin Nano
- remote TritonServer
- old implementation (remote only)

The native and containerized versions expose the metrics configuration, while the remote one hasn't any property. However, they still implement the `InferenceEngineMetricService` since the metrics are likely enabled in the Triton Server default configuration.

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 4]
```diff
@@ -158,6 +158,33 @@
...
             min="1">
         </AD>
+        
+        <AD id="enable.metrics"
+            name="Enable Triton Server Metrics"
+            type="Boolean"
+            cardinality="0"
+            required="true"
+            default="true"
+            description="Enable the Triton Server Metrics feature. This property enables the default, CPU and GPU metrics, if available.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Enable the Triton Server Metrics feature. This property enables the default CPU and GPU metrics, if available.">
```

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 5]
```diff
@@ -158,6 +158,33 @@
...
+        </AD>  
+        
+        <AD id="metrics.config"
+            name="Optional metrics-specific configuration"
+            type="String"
+            cardinality="0"
+            required="false"
+            default=""
+            description="A semi-colon separated list of metrics-specific configuration settings for the Triton Server Metrics. (i.e. counter_latencies=true)">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="A semi-colon separated list of Triton Server Metrics. (i.e. counter_latencies=true)">
```

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 6]
```diff
@@ -158,6 +158,33 @@
...
+        </AD>
+        
+        <AD id="metrics.interval"
+            name="Optional metrics polling interval"
+            type="String"
+            cardinality="0"
+            required="false"
+            default=""
+            description="Sets the interval in milliseconds at which the metrics are updated by the engine. This property is applied only to metrics that are designated as 'Per Interval'. Leave empty for the default value.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Sets the interval in milliseconds at which the engine updates the metrics. This property applies only to metrics designated as 'Per Interval.' Leave it empty to use the default value.">
```

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml (hunk 4 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]: @MMaiero The property enables the metrics feature that contains CPU, GPU and standard statistics. I'll rephrase in this way:

> Enable the Triton Server Metrics feature. This property enables the standard statistics and CPU/GPU metrics, if available.

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml (hunk 5 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]: @MMaiero This property allows to configure a metric, enabling or disabling the statistics that are emitted by the service. For example, when the metrics are enabled, a user can disable the `counter_latencies` setting `--metrics-config counter_latencies=false`. So, they aren't new metrics, but configurations for metrics that are already there.

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java [hunk 7]
```diff
@@ -264,6 +264,8 @@ private ContainerConfiguration createContainerConfiguration() {
         if (!this.options.getBackendsConfigs().isEmpty()) {
             this.options.getBackendsConfigs().forEach(config -> entrypointOverride.add("--backend-config=" + config));
         }
+        entrypointOverride.add("--allow-metrics=" + this.options.areMetricsEnabled());
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java:267]: Given that we're collecting model statistics from the [statistics extension](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protocol/extension_statistics.html) endpoint and GPU metrics from the [metric](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/metrics.html) one, is this still necessary?

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/metrics/parser/ModelStatisticsParser.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            ModelStatistics tritonModelStatistics = new ModelStatistics(statistic);
+            statistics.put(key, gson.toJson(tritonModelStatistics));
+        });
+
+        return statistics;
+    }
+
+    private String getKey(inference.GrpcService.ModelStatistics statistic) {
+        return "model.metrics." + statistic.getName() + "." + statistic.getVersion();
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/metrics/parser/ModelStatisticsParser.java]: Admittedly, this code is *really* hard to follow... I would suggest a couple of variable renames:

```suggestion
        Map<String, String> jsonStatistics = new HashMap<>();

        this.modelStatistics.forEach(grpcStatistic -> {
            String key = getKey(grpcStatistic);
            ModelStatistics tritonModelStatistics = new ModelStatistics(grpcStatistic);
            jsonStatistics.put(key, gson.toJson(tritonModelStatistics));
        });

        return jsonStatistics;
    }

    private String getKey(inference.GrpcService.ModelStatistics grpcStatistic) {
        return "model.metrics." + grpcStatistic.getName() + "." + grpcStatistic.getVersion();
    }
```

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/metrics/parser/ModelStatisticsParser.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        Map<String, String> statistics = new HashMap<>();
+
+        this.modelStatistics.forEach(statistic -> {
+            String key = getKey(statistic);
+            ModelStatistics tritonModelStatistics = new ModelStatistics(statistic);
+            statistics.put(key, gson.toJson(tritonModelStatistics));
+        });
+
+        return statistics;
+    }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/metrics/parser/ModelStatisticsParser.java]: Can we leverage the built-in grpc to JSON converter?

See:
- https://www.baeldung.com/java-convert-json-protobuf
- https://stackoverflow.com/a/42615002/7780889

**Review by mattdibi — CHANGES_REQUESTED** [review]

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java (hunk 7 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java:267]: The `--allow-metrics` is a general switch for the metrics. So, setting it to false and `--allow-gpu-metrics` to true causes to not receive anything.

Comment (mattdibi) [reply kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java:267]: Uh... not that intuitive I have to admit. Thanks for the explaination

**Comment by pierantoniomerlino** [comment]
The parser for the model statistics has been updated using the [JsonFormat](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/util/JsonFormat.html) class provided by `com.google.protobuf:protobuf-jav-util`. The library is embedded in the `org.eclipse.kura.ai.triton.server` bundle.

The library has been successfully checked using the Eclipse Dash License Tool:

```
echo "com.google.protobuf:protobuf-java-util:3.25.3" | java -jar org.eclipse.dash.licenses-1.1.1-20250220.065102-441.jar -
[main] INFO Querying Eclipse Foundation for license data for 1 items.
[main] INFO Found 0 items.
[main] INFO Querying ClearlyDefined for license data for 1 items.
[main] INFO Found 1 items.
[main] INFO Vetted license information was found for all content. No further investigation is required.
```

**Review by mattdibi — DISMISSED** [review]

**Review by mattdibi — APPROVED** [review]
